ISPN-7887 multimap hotrod, first simple implementation

karesti opened the pull request:
https://issues.jboss.org/browse/ISPN-7887

tristantarrant commented:
Temporarily closing because of CI overload

karesti commented:
@tristantarrant still todo moving the API to commons

karesti commented:
@tristantarrant @wburns I extracted a super interface that is in commons now. 
I didn't implement remove(k,v) and contains methods because I need first a validation on what I've already done. Then, adding those operations is quite straightforward and I can do it in another PR (or ticket) or even in this PR in a separate commit that I will amend just after.

rvansa commented on commons/src/main/java/org/infinispan/multimap/api/BasicMultimapCache.java:
```diff
@@ -0,0 +1,184 @@
+/**
+ * {@link BasicMultimapCache} provides the common building block for the two different types of multimap caches that Infinispan
+ * provides: embedded and remote. <p> Please see the <a href="http://infinispan.org/documentation/">Infinispan
+ * documentation</a> and/or the <a href="http://infinispan.org/docs/dev/getting_started/getting_started.html">5 Minute
```
Are those tutorials relevant?

karesti replied:
This are already merged on master in the multimap embedded doc

rvansa commented on commons/src/main/java/org/infinispan/multimap/api/BasicMultimapCache.java:
```diff
@@ -0,0 +1,184 @@
+ *
+ *    multimapCache.get("k").thenAccept(values -> System.out.println(values.size()));
+ * </pre>
+ * prints the value 2. "k" -> ["v1", "v"]
```
v -> v2

rvansa commented on commons/src/main/java/org/infinispan/multimap/api/BasicMultimapCache.java:
```diff
@@ -0,0 +1,184 @@
+ *  </pre>
+ * <p>
+ * <p>
+ * <h2>Duplicates</h2> The current implementation does not support duplicate values on keys. {@link
```
You should not talk about implementation in interface, probably, just say what duplicates are and link to factories that will clarify this.

karesti replied:
ok

rvansa commented on commons/src/main/java/org/infinispan/multimap/api/BasicMultimapCache.java:
```diff
@@ -0,0 +1,184 @@
+ * they are called in a explicit transaction context. This limitation could be improved in the following versions if
+ * technically possible.
+ * <p>
+ * <ul> <li>{@link BasicMultimapCache#size()} <li>{@link BasicMultimapCache#containsEntry(Object, Object)} </ul>
```
each item on new line please

rvansa commented on commons/src/main/java/org/infinispan/multimap/api/BasicMultimapCache.java:
```diff
@@ -0,0 +1,184 @@
+ * <p>
+ * Future implementations might evolve to support duplicated values.
+ * <p>
+ * <h2>Transactions</h2> MultimapCache supports implicit transactions without blocking. The following methods block when
```
ATM transactions are specific to embedded mode. Also, all Hot Rod operations are blocking.

karesti replied:
I think I have to review the doc of this super interface moved to commons

rvansa commented on commons/src/main/java/org/infinispan/multimap/api/package-info.java:
```diff
@@ -0,0 +1,6 @@
+ *
+ * @public
+ */
+package org.infinispan.multimap.api;
```
Do we need the `api` package? Why `o.i.multimap` is not sufficient?
Also since this is in commons, it should rather be `org.infinispan.commons.multimap[.api]`. I've seen some problems if multiple modules contain the same package.

karesti replied:
I'm following how it is done in Counters. Anyway, for me multimap and after api ... I would prefer org.infinispan.api.multimap, org.infinispan.api.counters, org.infinispan.api.locks etc ...

tristantarrant replied:
Yes, I'd prefer that too

rvansa commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/CacheDecodeContext.java:
```diff
@@ -322,44 +321,15 @@ Response get() {
-         EntryVersion entryVersion = ce.getMetadata().version();
-         long version = extractVersion(entryVersion);
-         byte[] v = ce.getValue();
-         int lifespan = ce.getLifespan() < 0 ? -1 : (int) ce.getLifespan() / 1000;
```
What happened to lifespan and maxIdle?

karesti replied:
I refactored the code

rvansa commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/Encoder2x.java:
```diff
@@ -331,7 +332,7 @@ public void writeResponse(Response response, ByteBuf buf, EmbeddedCacheManager c
             if (isTrace) log.trace("About to respond to bulk get request");
             if (r.status == OperationStatus.Success) {
-               try (CloseableIterator<Map.Entry<byte[], byte[]>> iterator = r.entries.iterator()) {
+               try (CloseableIterator<Map.Entry<byte[], byte[]>> iterator = (CloseableIterator<Map.Entry<byte[], byte[]>>) r.entries.iterator()) {
```
is this cast needed?

karesti replied:
hmmm it was probably not compiling for some reason in my laptop. Need to recheck

rvansa commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/HotRodHeader.java:
```diff
@@ -53,4 +53,8 @@ public String toString() {
             '}';
    }
+
+   public boolean isMultimap() {
```
-1 as per the flag critics above

karesti replied:
Yes, this is what I say, that I first used the flag to reuse operations but in the end I ended to create each operation, so, despite the fact that I create a context for multimap specifically on the server, os not useful anymore, so I just left it in case you guys see any additional use for the flag

karesti commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/Response.java:
```diff
@@ -305,19 +312,13 @@ public String toString() {
+                           OperationStatus status, int topologyId, CacheEntry<byte[], byte[]> ce) {
+      super(version, messageId, cacheName, clientIntel, operation, status, topologyId, ce == null ? null : ce.getValue());
+      this.dataVersion = extractVersion(ce);
+      this.created = extractCreated(ce);
```
@rvansa I refactored the code for the lifespan etc

karesti commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/MetadataUtils.java:
```diff
@@ -0,0 +1,49 @@
+import org.infinispan.container.versioning.NumericVersion;
+import org.infinispan.container.versioning.SimpleClusteredVersion;
+
+public class MetadataUtils {
```
@rvansa I created this class to handle the metadata directly from the entry instead calculating things in the other method

karesti commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/multimap/RemoteMultimapCacheManager.java:
```diff
@@ -0,0 +1,51 @@
+      this.remoteCacheManager = remoteCacheManager;
+   }
+
+   public <K, V> RemoteMultimapCache<K, V> get(String cacheName) {
```
I duplicated to see how it feels and if we want something like that, or not.

karesti commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/multimap/RemoteMultimapCache.java:
```diff
@@ -0,0 +1,22 @@
+ */
+public interface RemoteMultimapCache<K, V> extends BasicMultimapCache<K, V> {
+
+   CompletableFuture<MetadataValue<Collection<V>>> getWithMetadata(K key);
```
Could be a good idea. @tristantarrant WDYT ?

wburns replied:
Seems like an alright idea to me.

karesti replied:
ok

karesti commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/RemoteCacheManager.java:
```diff
@@ -347,17 +347,34 @@ public RemoteCacheManagerAdmin administration() {
+      return asyncExecutorService;
+   }
+
+   public static class RemoteCacheKey {
```
This is because I needed to use is in my other class when I separated the code (as fas as I remember). 
I want just to speak with the others which will be the best way to go here

karesti replied:
I need to review all this as you comment bellow, but internally when I create he cache I need a cache that is K Collection<V> and the createRemoteCache creates a K,V cache

karesti replied:
but again, i did this a long time ago and i juste duplicated some of the parts in purpose to see how to deal with everything after

karesti replied:
the hotrod client code is hard to enter at first

karesti replied:
I'm reviewing everything now

karesti commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/Flag.java:
```diff
@@ -54,7 +54,11 @@
+   /**
+    * Used by core to determine if the operation should be appliyed over a multimapcache or a regular one
+    */
+   IS_MULTIMAP(0x0020)
```
I do agree, seems useless at this point

wburns replied:
I agree, we should remove :+1:

wburns commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/multimap/operations/MultimapOperationsFactory.java:
```diff
@@ -0,0 +1,102 @@
+@Immutable
+public class MultimapOperationsFactory {
+
+   private final ThreadLocal<Integer> flagsMap = new ThreadLocal<>();
```
I personally am not a big fan of the way these flags worked in the original Client. I would say instead to store the flags on each instance of the RemoteMultiMap and pass them when they invoke something. When a new flag is added or removed we would create a new instance of the MultiMap. This seems much more intuitive to me and simpler.

Then the flags can be defaulted on the RemoteMultiMap when it is retrieved from the manager as well.

karesti replied:
Do you mind if I leave it like that, as the operations factory, and we try to do a global refactoring in another ticket for both ?

wburns replied:
Well I was thinking this way would do it properly and we could refactor the other later. But make sure a JIRA is created if you wanted to do this one later.

karesti replied:
ok

wburns commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/multimap/operations/RemoveMultimapOperation.java:
```diff
@@ -0,0 +1,36 @@
+         return Boolean.FALSE;
+      }
+
+      return transport.readVInt() == 1 ? Boolean.TRUE : Boolean.FALSE;
```
Why is this a VInt? Wouldn't we just return a single byte?

karesti commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/RemoteCacheManager.java:
```diff
@@ -318,10 +318,10 @@ private Properties loadFromStream(InputStream stream) {
    }
 
-   private void startRemoteCache(RemoteCacheHolder remoteCacheHolder) {
+   public void startRemoteCache(RemoteCacheHolder remoteCacheHolder) {
```
no

karesti replied:
no, I changed everything and removed duplicated code. At the time, I was unsure so I duplicated.

karesti commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/RemoteCacheManager.java:
```diff
@@ -347,17 +347,34 @@ public RemoteCacheManagerAdmin administration() {
    }
 
-   private static class RemoteCacheKey {
+   public ClientListenerNotifier getListenerNotifier() {
```
everything reviewed, all the duplicated code

karesti replied:
ok

karesti commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/RemoteCacheManager.java:
```diff
@@ -347,17 +347,34 @@ public RemoteCacheManagerAdmin administration() {
+      return listenerNotifier;
+   }
+
+   public ExecutorService getAsyncExecutorService() {
```
ok

karesti replied:
ok

karesti commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/multimap/RemoteMultimapCacheManagerFactory.java:
```diff
@@ -0,0 +1,19 @@
+   private RemoteMultimapCacheManagerFactory() {
+   }
+
+   public static RemoteMultimapCacheManager from(RemoteCacheManager remoteCacheManager) {
```
I don't get what you mean here

wburns replied:
He is saying that someone could just as easily call `new RemoteMultimapCacheManager(remoteCacheManager)` and not even use this factory. If we returned an interface there is no implementation details leaked.

karesti replied:
Ok

karesti replied:
I changed that to do it like in multimap embedded. I'm going to rework things to try to give the same API for remote and embedded. Just as a preview so we can discuss it

karesti replied:
(not in this PR)

karesti commented:
I just need to review java doc and the server side async calls

karesti commented:
and check OSGI fails

karesti commented on commons/src/main/java/org/infinispan/multimap/api/BasicMultimapCache.java:
```diff
@@ -0,0 +1,184 @@
+   CompletableFuture<Collection<V>> get(K key);
+
+   /**
+    * Asynchronous method. Removes the entry associated with the key from this multimap cache, if such exists.
```
ok

karesti commented:
@rvansa I've updated the comments. I'm just doing "join" on server side... is this for you ?

wburns commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/multimap/operations/ContainsValueMultimapOperation.java:
```diff
@@ -0,0 +1,78 @@
+      return transport.readByte() == 1 ? Boolean.TRUE : Boolean.FALSE;
+   }
+
+   //[header][key length][key][lifespan][max idle][value length][value]
```
This doesn't seem to write the key any longer.

karesti replied:
ok

wburns commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/multimap/metadata/MetadataCollectionImpl.java:
```diff
@@ -0,0 +1,71 @@
+   private final int maxIdle;
+   private final long version;
+
+   public MetadataCollectionImpl(Collection<V> collection, long created, int lifespan, long lastUsed, int maxIdle, long version) {
```
Maybe add another constructor that just takes the Collection. Also we should document on this class what time unit these values are assumed to be in. I am guessing MILLISECONDS.

karesti replied:
ok

wburns commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/multimap/RemoteMultimapCache.java:
```diff
@@ -0,0 +1,20 @@
+ */
+public interface RemoteMultimapCache<K, V> extends BasicMultimapCache<K, V> {
+
+   CompletableFuture<MetadataCollection<V>> getWithMetadata(K key);
```
I am thinking if we indeed want to return the metadata, maybe we should have a way to modify it. Do we want a `replaceWithVersion` or some way to override the default expiration parameters?

wburns replied:
Also please document the method.

karesti replied:
@wburns maybe, I'm going to log a ticket on that

karesti replied:
https://issues.jboss.org/browse/ISPN-8582

wburns commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/multimap/MetadataCollection.java:
```diff
@@ -0,0 +1,17 @@
+ */
+public interface MetadataCollection<V> extends Versioned, Metadata {
+
+   Collection<V> getCollection();
```
Missing javadoc as well.

karesti replied:
ok

karesti commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/multimap/RemoteMultimapCacheImpl.java:
```diff
@@ -0,0 +1,187 @@
+
+      PutKeyValueMultimapOperation<Void> op = operationsFactory.newPutKeyValueOperation(compatKey,
+            marshallKey, marshallValue, defaultLifespan, MILLISECONDS, defaultMaxIdleTime, MILLISECONDS);
+      return CompletableFuture.runAsync(() -> op.execute());
```
https://issues.jboss.org/browse/ISPN-8583

karesti commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/multimap/RemoteMultimapCacheImpl.java:
```diff
@@ -0,0 +1,187 @@
+
+   @Override
+   public CompletableFuture<Collection<V>> get(K key) {
+      assertRemoteCacheManagerIsStarted();
```
ok

karesti commented:
@wburns I addressed the comments, also @rvansa comment about the factory. It's in a separate commit I would like to amend when everything will be fine for you

karesti commented:
@wburns sure, I'll squash all the commits and relaunch the build

pruivo commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/LocalContextHandler.java:
```diff
@@ -25,12 +27,32 @@ public LocalContextHandler(NettyTransport transport) {
       if (msg instanceof CacheDecodeContext) {
          CacheDecodeContext cdc = (CacheDecodeContext) msg;
-         realChannelRead(ctx, msg, cdc);
+         if (cdc.header.op.isMultimap())
```
I have no idea why this code is here... and I have no idea why this class exists... /cc @wburns

wburns replied:
If I remember correctly this class is here to make operations that should not require blocking to be faster as they don't have to use a separate thread (in our case GET should always be local as we have smart routing for example). Let me confirm that though.

wburns replied:
Yeah that is what it is for https://github.com/karesti/infinispan/commit/a7526f4904f0f9e75f8e241f650f26820f1f41da

So in this case what she has for the Get should be fine.

So basically you want to add operations that you know won't block to this handler. And if it can block it should go in the standard [ContextHandler](https://github.com/infinispan/infinispan/blob/master/server/hotrod/src/main/java/org/infinispan/server/hotrod/ContextHandler.java#L61)

karesti replied:
@wburns I might need to change to use ReadEntryView instead of ReadWriteEntryView or we don't care about that ?

karesti replied:
@wburns the this is that I'm using the ReadWriteMap functional interface, even when I call just "get" (I don't change any value, but I'm calling that). Can this harm or the is no link at all ?

wburns replied:
Not sure I understand. As long as the operation isn't going to block this thread we can leave it here. The object is all stored in 1 node and it should be this node so any read should be fine.  It shouldn't matter that it is a write, the problem is that a write normally has to go remote (for example a LOCAL only cache in hotrod could use this handler as well)

pruivo commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/multimap/metadata/MetadataCollectionImpl.java:
```diff
@@ -0,0 +1,83 @@
+      b.append(", lifespan=");
+      b.append(", maxIdle=");
+      b.append(", collection=[");
+      b.append(collection.stream()
```
oh... you are copying Radim's `StringBuilder` code... what is wrong with `b.append(collection)`?

karesti replied:
Nothing, I didn't think further tbh

pruivo commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/multimap/metadata/MetadataCollectionImpl.java:
```diff
@@ -0,0 +1,83 @@
+   public String toString() {
+      StringBuilder b = new StringBuilder();
+      b.append("MetadataCollectionImpl{");
+      b.append("version=");
```
hmmm isn't an `append(version)` missing? the same for other fields.

karesti replied:
yes, I just forgot all the fields haha

pruivo commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/multimap/protocol/MultimapHeaderParams.java:
```diff
@@ -0,0 +1,33 @@
+import org.infinispan.client.hotrod.impl.protocol.HeaderParams;
+
+/**
+ *
```
missing javadoc

karesti replied:
ok

pruivo commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/multimap/operations/MultimapOperationsFactory.java:
```diff
@@ -0,0 +1,119 @@
+@Immutable
+public class MultimapOperationsFactory {
+
+   private final ThreadLocal<Integer> flagsMap = new ThreadLocal<>();
```
I don't see any usage of the thread local. IMO, it can be removed.

karesti replied:
Flags are passed to the operations through them ...

pruivo replied:
but I didn't find any `flagsMap.set()`. its value will be always zero (or null)

karesti replied:
you are right, I didn't realise before. I'm going to just rework all the flag stuff in the issue I opened after will's commets on this : https://issues.jboss.org/browse/ISPN-8580

pruivo commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/multimap/MultimapResponseHandler.java:
```diff
@@ -0,0 +1,84 @@
+
+import io.netty.buffer.ByteBuf;
+
+public final class MultimapResponseHandler {
```
javadoc?

karesti replied:
ok

pruivo commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/Decoder2x.java:
```diff
@@ -241,6 +241,7 @@ private Response createResponse(HotRodHeader h, OperationStatus st, byte[] prev)
             case Success:
                switch (h.op) {
                   case PUT:
+                  case PUT_MULTIMAP:
```
I don't think it belongs here.

karesti replied:
this is  a merge error

karesti replied:
not even a merge error ! I put that there a lot time ago, nice catch

pruivo commented on client/hotrod-client/pom.xml:
```diff
@@ -120,7 +120,13 @@
       </dependency>
 
-      <dependency>
+       <dependency>
```
duplicated dependency. it already depends on commons (see above, first dependency)

karesti replied:
I have no idea how this landed there....

pruivo commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/multimap/RemoteMultimapCacheImpl.java:
```diff
@@ -0,0 +1,195 @@
+
+      PutKeyValueMultimapOperation<Void> op = operationsFactory.newPutKeyValueOperation(compatKey,
+            marshallKey, marshallValue, defaultLifespan, MILLISECONDS, defaultMaxIdleTime, MILLISECONDS);
+      return CompletableFuture.runAsync(() -> op.execute());
```
you have an executor used for async operation in RemoteCacheManager you should use it!

karesti replied:
ok !

karesti replied:
I said Ok but I actually created an issue to deal that

https://issues.jboss.org/browse/ISPN-8583

karesti commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/multimap/operations/ContainsValueMultimapOperation.java:
```diff
@@ -0,0 +1,78 @@
+ * @author Katia Aresti, karesti@redhat.com
+ * @since 9.2
+ */
+public class ContainsValueMultimapOperation extends RetryOnFailureOperation<Boolean> {
```
because there is no key

karesti commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/multimap/operations/ContainsValueMultimapOperation.java:
```diff
@@ -0,0 +1,78 @@
+import org.infinispan.client.hotrod.impl.transport.TransportFactory;
+
+/**
+ * Implements "contains value" for multimap cache as defined by  <a href="http://community.jboss.org/wiki/HotRodProtocol">Hot
```
I didn't think about operation status, more than is a success and after sending the operation result which is a boolean in this case. which is your suggestion ? to send which operation status value ?

pruivo replied:
I use status `0x00` (Success)  to represent true, and `0x01` (OperationNotExecuted) to represent false.

karesti replied:
But the operation was executed...

karesti replied:
I think I'm gonna leave that like that for now, because fo me the semantics of operation not executed do not mean the same thing

pruivo replied:
I think you are overthinking too much :) it is just a byte. is there any case where the operation isn't executed?
But if you want to be consistent with Cache.containsKey, you have the status `0x02` (KeyDoesNotExist)

karesti commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/HotRodHeader.java:
```diff
@@ -53,4 +53,5 @@ public String toString() {
             ", topologyId=" + topologyId +
             '}';
    }
+
```
I have something there before and I remove it but the espace was left there

karesti commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/CrashedMemberDetectorListener.java:
```diff
@@ -27,7 +27,7 @@
 
    CrashedMemberDetectorListener(Cache<Address, ServerAddress> cache, HotRodServer server) {
-      // Let all nodes remove the address from their own cache locally.
+      // Let all nodes removeKey the address from their own cache locally.
```
ok

karesti commented on multimap/src/main/java/org/infinispan/multimap/impl/EmbeddedMultimapCache.java:
```diff
@@ -30,6 +31,33 @@
+ * they are called in a explicit transaction context. This limitation could be improved in the following versions if
+ * technically possible.
+ *
+ * <ul> <li>{@link BasicMultimapCache#size()} <li>{@link BasicMultimapCache#containsEntry(Object, Object)} </ul>
```
There is a formatting error. It means that the methods size and containsEntry block today on explicit tx contexts

karesti commented on commons/src/main/java/org/infinispan/multimap/api/BasicMultimapCache.java:
```diff
@@ -0,0 +1,160 @@
+    * <p>
+    * Returns duplicates are supported or not in this multimap cache.
+    *
+    * @return {@code true} if this multicache supports duplicate values for a given key.
```
multimap, thank you

karesti commented on commons/src/main/java/org/infinispan/multimap/api/BasicMultimapCache.java:
```diff
@@ -0,0 +1,160 @@
+   /**
+    * Returns {@link Boolean#TRUE} if this multimap cache contains the key.
+    *
+    * @param key the key tht might exists in this multimap cache
```
ok

karesti commented on commons/src/main/java/org/infinispan/multimap/api/BasicMultimapCache.java:
```diff
@@ -0,0 +1,160 @@
+   CompletableFuture<Collection<V>> get(K key);
+
+   /**
+    * Removes all the key-valu pairs associated with the key from this multimap cache, if such exists.
```
ok

karesti commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/multimap/operations/GetKeyMultimapOperation.java:
```diff
@@ -0,0 +1,58 @@
+   }
+
+   @Override
+   protected Transport getTransport(int retryCount, Set<SocketAddress> failedServers) {
```
ok

karesti commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/multimap/operations/ContainsEntryMultimapOperation.java:
```diff
@@ -0,0 +1,39 @@
+import net.jcip.annotations.Immutable;
+
+/**
+ * Implements "contains entry" for multimap as defined by  <a href="http://community.jboss.org/wiki/HotRodProtocol">Hot
```
ok

karesti commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/multimap/operations/AbstractMultimapKeyValueOperation.java:
```diff
@@ -0,0 +1,27 @@
+import org.infinispan.client.hotrod.impl.protocol.HeaderParams;
+import org.infinispan.client.hotrod.impl.transport.TransportFactory;
+
+public abstract class AbstractMultimapKeyValueOperation<T> extends AbstractKeyValueOperation<T> {
```
ok

karesti commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/multimap/operations/AbstractMultimapKeyOperation.java:
```diff
@@ -0,0 +1,22 @@
+import org.infinispan.client.hotrod.impl.protocol.HeaderParams;
+import org.infinispan.client.hotrod.impl.transport.TransportFactory;
+
+public abstract class AbstractMultimapKeyOperation<V> extends AbstractKeyOperation<V> {
```
ok

karesti commented on client/hotrod-client/src/main/java/org/infinispan/client/hotrod/Flag.java:
```diff
@@ -54,7 +54,7 @@
     * Used by the Query module only, it will prevent the indexes to be updated as a result of the current operations.
     */
-   SKIP_INDEXING(0x0010)
+   SKIP_INDEXING(0x0010),
```
ok

karesti commented on server/hotrod/src/main/java/org/infinispan/server/hotrod/HotRodOperation.java:
```diff
@@ -73,6 +73,17 @@
    COUNTER_GET_NAMES(0x64, 0x65, EnumSet.of(OpReqs.REQUIRES_AUTH), DecoderRequirements.HEADER_CUSTOM),
 
+   // Multimap operations
+   GET_MULTIMAP(0x67, 0x68, EnumSet.of(OpReqs.REQUIRES_KEY, OpReqs.REQUIRES_AUTH, OpReqs.CAN_SKIP_CACHE_LOAD, OpReqs.IS_MULTIMAP), DecoderRequirements.KEY),
```
We were using the same numbers I had conflicts on that !

karesti replied:
In any case, at one point this should evolve because we are getting rig of bytes and response codes from server to client are handled as shorts

https://issues.jboss.org/browse/ISPN-8590

karesti commented:
@pruivo comments addressed

karesti commented:
@pruivo I understand now what the thing in the pom is, I have added the "optional" in the commons dependency but I don't think this is what it should be done... I have karaf compilation problems in remote cache ... nvm, I'm going to add the changed required label even if i want to test what this is doing

karesti commented:
@pruivo as far as I'm concerned now, I think I'm going to leave it like that the PR and rework it just after

tristantarrant commented on multimap/src/main/java/org/infinispan/multimap/api/package-info.java:
```diff
@@ -0,0 +1,6 @@
+ *
+ * @public
+ */
+package org.infinispan.multimap.api;
```
This package should only be present in the commons Jar. Additional API "extensions" for embedded and remote should be further qualified. We must not have classes present in the same package in different jars.

karesti replied:
is it ok if I moved them to api.embedded ?

karesti replied:
so I have package org.infinispan.multimap.api.embedded;

karesti commented on multimap/src/main/java/org/infinispan/multimap/api/embedded/EmbeddedMultimapCacheManagerFactory.java:
```diff
@@ -1,4 +1,4 @@
-package org.infinispan.multimap.api;
+package org.infinispan.multimap.api.embedded;
```
@tristantarrant If you are ok on the change and so does jenkins, I would like to squash the commits and merge

karesti commented:
Failing tests are not related ...

ryanemerson commented:
Integrated, thanks @karesti
